docs/learn.md

```diff
@@ -721,20 +723,1250 @@ template](https://github.com/melange-re/melange-opam-template).
+on top of an OCaml language construct called `external`.
+
+`external` is a keyword for declaring a value in OCaml that will [interface with
+C code](https://v2.ocaml.org/manual/intfc.html):
+
+```ocaml
+external my_c_function : int -> string = "someCFunctionName"
+```
+
+It is like a `let` binding, except that the body of an external is a string.
+That string has a specific meaning depending on the context. For native OCaml,
+it usually refers to a C function with that name. For Melange, it refers to the
+functions or values that exist in the runtime JavaScript code, and will be used
+from Melange.
+
+Melange externals are always decorated with certain `[@bs.xxx]` attributes.
+These attributes are listed [above](#list-of-attributes-and-extensions), and
+will be further explain in the next sections.
+
+Once declared, one can use an `external` as a normal value.
+
+Melange external functions are turned into the expected JavaScript values,
+inlined into their callers during compilation, and completely erased afterwards.
+They don’t appear in the JavaScript output, so there are no costs on bundling
+size.
+
+**Note**: it is recommended to use external functions and the `[@bs.xxx]`
+attributes in the interface files as well, as this allows some optimizations
+where the resulting JavaScript values can be directly inlined at the call sites.
+
+#### Special identity external
+
+One external worth mentioning is the following one:
+
+```ocaml
+type foo = string
+type bar = int
+external danger_zone : foo -> bar = "%identity"
+```
+
+This is a final escape hatch which does nothing but convert from the type `foo`
+to `bar`. In the following sections, if you ever fail to write an `external`,
+you can fall back to using this one. But try not to.
+
+Let’s now see all the ways to use JavaScript from Melange.
+
+### Generate raw JavaScript
+
+It is possible to directly write JavaScript code from a Melange file. This is
+unsafe, but it can be useful for prototyping or as a escape hatch.
+
+To do it, we will use the `bs.raw`
+[extension](https://v2.ocaml.org/manual/extensionnodes.html):
+
+```ocaml
+let add = [%bs.raw {|
+  function(a, b) {
+    console.log("hello from raw JavaScript!");
+    return a + b;
+  }
+|}]
+
+let () = Js.log (add 1 2)
+```
+
+The `{||}` strings are called ["quoted
+strings"](https://ocaml.org/manual/lex.html#quoted-string-id). They are similar
+to JavaScript’s template literals, in the sense that they are multi-line, and
+there is no need to escape characters inside the string.
+
+Using one `%` (`[%bs.raw <string>]`) is useful to define expressions (function
+bodies, or other values) where the implementation is directly JavaScript. This
+is useful as we can attach the type signature already in the same line, to make
+our code safer. For example:
+
+```ocaml
+[%%bs.raw "var a = 1"]
+
+let f : unit -> int = [%bs.raw "function() {return 1}"]
+```
+
+Using two percentages `%%` (`[%%bs.raw <string>]`) is reserved for top-level
```

Comment:
```suggestion
Using two percentage signs (`[%%bs.raw <string>]`) is reserved for top-level
```

Reply:
not just top-level definitions, but rather structure items, e.g. 

```ocaml

module Foo = struct
  [%%bs.raw "not really toplevel"]
end
```

Reply:
I did not include this particular example for simplicity. But removed the "top-level" reference, mentioned structure and signature definitions, and linked to OCaml manual in https://github.com/melange-re/documentation-site/pull/12/commits/787e2619e9ca835178c61918f01b9ad45ed97eac.